docs/BuildingLBANN.md

```diff
@@ -0,0 +1,382 @@
+  NVTX. NVTX is provided as part of the CUDA toolkit.
++ VTune. LBANN supports some improved annotations for VTune.
+
+The following LLNL-maintained packages are required.
+
++ [Hydrogen](https://github.com/llnl/elemental) is a fork of the
+  Elemental distributed dense linear-algebra library and it may be
+  installed via [spack](https://github.com/llnl/spack) using the
+  package name "hydrogen". If CUDA support is required in LBANN,
+  Hydrogen must also be build with CUDA support.
+
+The following LLNL-maintained packages are optional.
+
++ [Aluminum](https://github.com/llnl/aluminum) is a
+  communication library optimized for machine learning and interaction
+  with GPUs. We cannot recommend its use strongly enough. It can be
+  built using [spack](https://github.com/llnl/spack).
++ [CONDUIT](https://github.com/llnl/conduit) is used to ingest
+  structured data produced by scientific simulations.
+
+
+## Building with [Spack](https://github.com/llnl/spack)
+
+Some variation on the theme of `spack install lbann`. bvanessen should
+document the spack flags or whatever they call them that LBANN
+supports. I don't know the current state of spack with respect to its
+building LBANN correctly and/or successfully.
+
+## Buidling with [CMake](https://cmake.org)
+
+LBANN uses [CMake](https://cmake.org) for its build system and a
+version newer than or equal to 3.9.0 is required. LBANN development is
+done primarily on UNIX-based platforms. As such, the build is tested
+regularly on Linux-based machines, occasionally on OSX, and never on
+Windows machines.
+
+It is required that LBANN be built out-of-source. That is, CMake must
+not be invoked in a directory containing a CMakeLists.
+
+### LBANN CMake options
+The following options are exposed in the CMake build system.
+
++ `LBANN_WITH_ALUMINUM` (Default: `OFF`): Use the Aluminum communication
+  package. This will be set to `ON` automatically if Hydrogen was
+  built with Aluminum.
+
++ `LBANN_WITH_CNPY` (Default: `ON`): Build with support for CNPY for reading
+  Numpy data.
+
++ `LBANN_WITH_CUDA` (Default: `OFF`): Enable a CUDA-aware build. If
+  this flag is `ON` and Hydrogen CUDA support cannot be detected,
+  LBANN's configuration step will fail.
+
++ `LBANN_WITH_CONDUIT` (Default: `OFF`): Build with support for CONDUIT.
+
++ `LBANN_WITH_NVPROF` (Default: `OFF`): Build with extra annotations for NVPROF.
+
++ `LBANN_WITH_TOPO_AWARE` (Default: `ON`): Use HWLOC for topology-aware choices.
+
++ `LBANN_WITH_TBINF` (Default: `ON`): Enable the Tensorboard interace.
+
++ `LBANN_WITH_VTUNE` (Default: `OFF`): Build with extra annotations for VTune.
+
++ `LBANN_DETERMINISTIC` (Default: `OFF`): Force as much of the code as possible
+  to be deterministic. This is not a guarantee as certain operations
+  in third-party libraries cannot be forced into a deterministic mode,
+  especially for CUDA-enabled builds.
+
++ `LBANN_SEQUENTIAL_INITIALIZATION` (Default: `OFF`): Force sequentially
```

Comment:
We should document this better, because we're not really referring to [sequential consistency](https://en.wikipedia.org/wiki/Consistency_model#Sequential_Consistency), but that initialization is the same regardless of how many processors participate. (I think I'm the one who actually called it "sequential consistency", but that doesn't mean we should not embellish upon my poor choices.)

Reply:
I'm not sure if we actually use `LBANN_SEQUENTIAL_INITIALIZATION` anywhere. I think that functionality was folded into `LBANN_DETERMINISTIC` in PR #363.